src/_P121_HMC5883L.ino

```diff
@@ -0,0 +1,154 @@
+{
+  boolean success = false;
+
+  switch (function)
+  {
+  case PLUGIN_DEVICE_ADD:
+  {
+    Device[++deviceCount].Number = PLUGIN_ID_121;
+    Device[deviceCount].Type = DEVICE_TYPE_I2C;
+    Device[deviceCount].VType = Sensor_VType::SENSOR_TYPE_QUAD;
+    Device[deviceCount].Ports = 0;
+    Device[deviceCount].PullUpOption = false;
+    Device[deviceCount].InverseLogicOption = false;
+    Device[deviceCount].FormulaOption = true;
+    Device[deviceCount].ValueCount = 4;
+    Device[deviceCount].SendDataOption = true;
+    Device[deviceCount].TimerOption = true;
+    Device[deviceCount].GlobalSyncOption = true;
+    success = true;
+    break;
+  }
+
+  case PLUGIN_GET_DEVICENAME:
+  {
+    string = F(PLUGIN_NAME_121);
+    success = true;
+    break;
+  }
+
+  case PLUGIN_GET_DEVICEVALUENAMES:
+  {
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_121));
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], PSTR(PLUGIN_VALUENAME2_121));
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[2], PSTR(PLUGIN_VALUENAME3_121));
+    strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[3], PSTR(PLUGIN_VALUENAME4_121));
+    success = true;
+    break;
+  }
+  case PLUGIN_WEBFORM_SHOW_I2C_PARAMS:
+  {
+    success = true;
+    break;
+  }
+  case PLUGIN_I2C_HAS_ADDRESS:
+  {
+    success = (event->Par1 == 0x1E);
+    break;
+  }
+  case PLUGIN_WEBFORM_LOAD:
+  {
+    addFormFloatNumberBox(F("DeclinationAngle"), F("plugin_121_HMC5883L_decl"), PCONFIG_FLOAT(0), -2.0f,2.0f ,2, 0.01f);
+    addUnit(F("degree"));
+    success = true;
+    break;
+  }
+
+  case PLUGIN_WEBFORM_SAVE:
+  {
+    PCONFIG_FLOAT(0) = getFormItemFloat(F("plugin_121_HMC5883L_decl"));
+    success = true;
+    break;
+  }
+
+  case PLUGIN_INIT:
+  {
+    initPluginTaskData(event->TaskIndex, new (std::nothrow) P121_data_struct());
+    P121_data_struct *P121_data =
+        static_cast<P121_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+    if (nullptr != P121_data)
+    {
+      P121_data->begin(event->TaskIndex);
+      success = P121_data->initialized;
+    }
+    break;
+  }
+  case PLUGIN_READ:
+  {
+    P121_data_struct *P121_data =
+        static_cast<P121_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+    if (nullptr != P121_data)
+    {
+      P121_data->begin(event->TaskIndex);
+
+      if (!P121_data->initialized)
+      {
+        addLog(LOG_LEVEL_ERROR, F("Could not initialize HCM5883L"));
+        break;
+      }
+      sensors_event_t s_event;
+      P121_data->mag.getEvent(&s_event);
+      UserVar[event->BaseVarIndex + 0] = s_event.magnetic.x;
+      UserVar[event->BaseVarIndex + 1] = s_event.magnetic.y;
+      UserVar[event->BaseVarIndex + 2] = s_event.magnetic.z;
+
+      float heading = atan2(s_event.magnetic.y, s_event.magnetic.x);
+
+      const float decl = PCONFIG_FLOAT(0);
+      
+      if (decl != 0)
+      {
+        heading += decl;
+      }
+
+      if (heading < 0)
```

Comment:
Don't expect that `float` comparison is going to be very reliable (though I'm not an expert on C++ float handling, I do know it should probably handled by one of the support functions from `ESPEasy_Math` helper, like `definitelyLessThan` and `essentiallyEqual`)

Reply:
I would like to request help, since I am not a cpp expert. If someone can correct this, I would appreciate it.

Reply:
I also learned most of that stuff from comments on my code 😆

Reply:
Just search through the code for the functions defined in the ESPEasy_math helper file :)
The simple idea is that comparing float values can yield unexpected results, since not all floating point values can be exactly represented in a float.
On top of that, the registers in the math (co)processor of an i386 or newer CPU to store a floating point value have slightly higher nr of bits available.
This makes comparing floats even more unpredictable as it depends on whether values are still in the registers or offloaded to RAM in between.

In essence you need to compare floating point values with some margin of error.
Thus consider 2 floats being equal if their  (absolute) difference is less than some value epsilon.
Epsilon depends on the values you try to compare too, but let's assume we are comparing float values roughly equal to 1.
Epsilon is then the smallest value where: (1+epsilon) != 1
Thus consider 2 floats equal when: abs(a - b) < epsilon.

For floats larger than 1, you simply multiply epsilon by the largest (absolute) value of both floats to compare.

That's the very short introduction to numerical math. 

Donald Knuth has written a number of books on this subject. 
Fun fact: You can even "earn" some money if you can find errors in his books. He writes a cheque to everyone who finds an error and for every next error you find the amount in dollars doubles.
It is estimated he has written cheques for a total  amount of more than 1 million $, but I doubt anyone will every cash such a cheque as it is quite a status symbol to have one :)